File: algo/acm/leetcode/wordladder.cc
#include <iostream>
#include <unordered_set>
#include <map>
#include <string>
#include <queue>

using namespace std;


int bfs(const string &start, const string &end,
        map<string,int> &dis,
        unordered_set<string> &visited,
        unordered_set<string> &dict
        ) {
  queue<string> q;
  q.push(start);
  visited.insert(start);
  dis[start] = 1;
  while (!q.empty()) {
    string u = q.front();
    q.pop();
    string v = u;
    for (int i = 0; i < u.length(); i++) {
      for (char c = 'a'; c <= 'z'; c+=1) {
        if (c != v[i]) {
          char tmp = v[i];
          v[i] = c;
          if (dict.find(v) != dict.end() &&
              visited.find(v) == visited.end()) {
            dis[v] = dis[u] + 1;
            if (v == end) return dis[v];
            visited.insert(v);
            q.push(v);
          }
          v[i] = tmp;
        }
      }
    }
  }
  return 0;
}

class Solution {
public:
    int ladderLength(string start, string end, unordered_set<string> &dict) {
      unordered_set<string> visited;
      map<string, int> dis;
      dict.insert(start);
      dict.insert(end);
      if (start == end) return 1;
      return bfs(start, end, dis, visited, dict);
    }
};


int main(int argc, const char *argv[])
{

  string start = "hit";
  string end = "hot";
  unordered_set<string> dict({"hot","dot","dog","lot","log"});
  Solution sol;
  cout << sol.ladderLength(start, end, dict) << endl;
  return 0;
}

File: algo/acm/leetcode/wordladder.py
import collections

alphabets = ['a',
             'b',
             'c',
             'd',
             'e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']

def makeGraph(d):
    def connectedIter(s):
        s_list = list(s)
        for i in xrange(0, len(s_list)):
            for c in alphabets:
                if c != s[i]:
                    s_list[i] = c;
                    tmp_s = ''.join(s_list)
                    if tmp_s in d:
                        yield tmp_s
                    s_list[i] = s[i]
    return connectedIter

def bfs(start, end, dis, visited, g):
    q = collections.deque()
    q.append(start)
    dis[start] = 1
    visited.add(start)
    while q:
        u = q.popleft()
        dis_u = dis[u]
        for v in g(u):
            if v not in visited:
                if v == end:
                    return dis_u + 1
                q.append(v)
                visited.add(v)
                dis[v] = dis_u + 1
    return 0


class Solution:
    # @param start, a string
    # @param end, a string
    # @param dict, a set of string
    # @return an integer
    def ladderLength(self, start, end, d):
        visited = set()
        dis = dict()
        d.add(end)
        d.add(start)
        return bfs(start, end, dis, visited, makeGraph(d))
